lib/gateway/NodeGateway.ts

```diff
@@ -1,122 +1,44 @@
+  LocalGraphQLDataSource,
 } from '@apollo/gateway';
-import { ServiceDefinition } from '@apollo/federation';
-import { SchemaDirectiveVisitor } from 'graphql-tools';
-import gql from 'graphql-tag';
-import {
-  DocumentNode,
-  graphql,
-  GraphQLSchema,
-  ObjectTypeDefinitionNode,
-  parse,
-  visit,
-} from 'graphql';
-
-import createNodeService from '../service';
-import TypeIDDataSource from '../utils/TypeIDDataSource';
 
-interface NodeGatewayConfig {
-  typeIDDataSource: TypeIDDataSource;
-  directiveResolvers?: {
-    [name: string]: typeof SchemaDirectiveVisitor;
-  };
-  directiveTypeDefs?: DocumentNode | DocumentNode[];
-  nodeQueryDirectives?: string | null;
-}
+import IntrospectAddNodeAndCompose from 'lib/subgraphManagers/IntrospectAddNodeAndCompose';
+import { NODE_SERVICE_NAME, NODE_SERVICE_URL } from 'lib/utils/constants';
 
 export default class NodeGateway extends ApolloGateway {
-  private static isNodeInterface = (node: ObjectTypeDefinitionNode): boolean =>
-    node.interfaces?.some(({ name }): boolean => name.value === 'Node') ||
-    false;
-
-  private static generateNodeType = (name: string): DocumentNode =>
-    gql`
-    extend type ${name} implements Node @key(fields: "id") {
-      id: ID! @external
-    }
-  `;
-
-  private nodeGatewayConfig: NodeGatewayConfig;
+  private getNodeServiceDataSource: () => LocalGraphQLDataSource;
 
-  constructor(nodeGatewayConfig: NodeGatewayConfig, config?: GatewayConfig) {
+  constructor(config?: GatewayConfig) {
     const { buildService: oldBuildService } = config || {};
     const buildService = (
-      serviceDef: ServiceEndpointDefinition & { schema?: GraphQLSchema },
+      serviceDef: ServiceEndpointDefinition,
     ): GraphQLDataSource => {
-      const { schema } = serviceDef;
-      if (schema) return new LocalGraphQLDataSource(schema);
-      return oldBuildService
-        ? oldBuildService(serviceDef)
-        : new RemoteGraphQLDataSource(serviceDef);
+      const { name, url } = serviceDef;
+      let dataSource: GraphQLDataSource | undefined;
+      if (name === NODE_SERVICE_NAME && url === NODE_SERVICE_URL) {
+        dataSource = this.getNodeServiceDataSource?.();
+      }
+      if (!dataSource) {
+        dataSource = oldBuildService
+          ? oldBuildService(serviceDef)
+          : new RemoteGraphQLDataSource(serviceDef);
+      }
+      return dataSource;
     };
     super({ ...config, buildService });
-    this.nodeGatewayConfig = nodeGatewayConfig;
-  }
-
-  protected async loadServiceDefinitions(
-    config: GatewayConfig,
-  ): ReturnType<ApolloGateway['loadServiceDefinitions']> {
-    const defs = await super.loadServiceDefinitions(config);
-    const nodeTypeDefs: DocumentNode[] = [];
-    const seenNodeTypes = new Set<string>();
-    if (!defs.serviceDefinitions) {
-      return defs;
+    if (
+      config &&
+      'supergraphSdl' in config &&
+      config.supergraphSdl instanceof IntrospectAddNodeAndCompose
+    ) {
```

Comment:
I'd simplify as:

```ts
const { supergraphSdl } = config || {};
if (supergraphSdl instanceof IntrospectAddNodeAndCompose) {
 ...
```

actually merging this into the other suggested code, since we'd not need to declare `getNodeServiceDataSource`, rewriting my previous suggestion:

```ts
const getNodeServiceDataSource = (config: GatewayConfig, serviceDef: ServiceEndpointDefinition): GraphQLDataSource | undefined => {
      const { name, url } = serviceDef; // any way to get this from `service`?
      if (name === NODE_SERVICE_NAME && url === NODE_SERVICE_URL) {
        const { supergraphSdl } = config || {};
        if (supergraphSdl instanceof IntrospectAddNodeAndCompose) {
           return supergraphSdl.nodeServiceLocalGraphQLDataSource();
        }
      }
      return undefined;
}
```

which is cleaner